app/controllers/stats_controller.rb

```diff
@@ -2,29 +2,42 @@ class StatsController < ApplicationController
 
   before_action :authorize_user
 
+  SPREADSHEET_CONTENT_TYPE = 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
+
   def index
     @foi_reports = ReportType.standard.foi.order(:full_name)
     @sar_reports = ReportType.standard.sar.order(:full_name)
   end
 
   def download
     report = Report.new report_type_id: params[:id]
-    report.run
 
-    send_data report.report_data, filename: report.report_type.filename
+    if report.xlsx?
+      report_data = report.run
+
+      axlsx = create_spreadsheet(report_data)
+
+      send_data axlsx.to_stream.read,
+                filename: report.report_type.filename('xlsx'),
+                disposition: :attachment,
+                type: SPREADSHEET_CONTENT_TYPE
+    else
+      report.run_and_update!
+      report.trim_older_reports
+
+      send_data report.report_data, filename: report.report_type.filename('csv')
+    end
   end
```

Comment:
The `if report.xlsx? ... else ... end` condition suggests this chunk of code should be inside the `Report` object or another place perhaps, or maybe 2 classes: `ExcelReport` and `CSVReport` ? Feel uneasy about this block being in controller

Reply:
report shouldn't be a model - its more like an 'action'. Felt this was the best solution rather than a complete re-write.